src/protections/NsfwProtection.ts

```diff
@@ -17,20 +17,113 @@ limitations under the License.
 import { Protection } from "./IProtection";
 import { Mjolnir } from "../Mjolnir";
 import * as nsfw from "nsfwjs";
-import { LogLevel, LogService } from "@vector-im/matrix-bot-sdk";
-import { node } from "@tensorflow/tfjs-node";
+import { LogLevel, LogService, MXCUrl } from "@vector-im/matrix-bot-sdk";
+import { node, Tensor3D } from "@tensorflow/tfjs-node";
+import { spawn } from "node:child_process";
+import { ReadableStream } from "node:stream/web";
+
+const TENSOR_SUPPORTED_TYPES = [
+    "image/png",
+    "image/apng",
+    "image/bmp",
+    "image/x-bmp",
+    "image/gif",
+    "image/jpeg",
+    "image/jp2",
+    "image/jpx",
+    "image/jpm",
+];
+
+const FFMPEG_SUPPORTED_TYPES = ["video/", "image/"];
+
+const FFMPEG_EXTRA_ARGS: Record<string, string[]> = {
+    // Extra params needed to extract svg thumbnails.
+    "image/svg+xml": ["-f", "svg_pipe", "-frame_size", "10000", "-video_size", "512x512"],
+};
 
 export class NsfwProtection extends Protection {
     settings = {};
     // @ts-ignore
-    private model: any;
+    private model: nsfw.NSFWJS;
+
+    private classificationCache = new Map<string, boolean>();
+
+    /**
+     * Extract the first frame from a video or image source.
+     * @param ffmpegPath The path to the `ffmpeg` binary.
```

Comment:
Should it pull from somewhere in the first third-middle? Just wondering if always pulling from first frame will be easily evaded.